b2: Add b2-version-at flag to show file versions at time

swazrgb opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

The goal is to be able to browse the b2 bucket state as it was at a specified point in time, using b2's file versions. This is especially useful in combination with other backup tools, such as restic, which may use rclone as a backend.

I have not added any tests, since I did not find any for `--b2-versions` which is similar functionality but I might be looking in the wrong place.

For example, this allows the following:

```
rclone mount --b2-version-at=2022-03-13T00:00:00Z b2:bucket /mnt/rclone
```

Or when using a backup tool such as restic:

```
restic -o 'rclone.args=serve restic --stdio --b2-version-at=2022-03-13T00:00:00Z' --no-lock mount /mnt/restic
```

To access pruned backup files whose data is still kept by b2 lifecycle rules.

#### Was the change discussed in an issue or in the forum before?

Yes, but not by me: https://github.com/rclone/rclone/issues/2126

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on backend/b2/b2.go:
```diff
@@ -696,9 +703,20 @@ func (f *Fs) list(ctx context.Context, bucket, directory, prefix string, addBuck
 		opts.Path = "/b2_list_file_versions"
 	}
+
+	var versionAtTime time.Time
```
This should be parsed into a time.Time in NewFs so we only do it once at the start.

We can then use `f.versionAtTime.IsZero()` to see if the feature is active.

ncw commented on backend/b2/b2.go:
```diff
@@ -662,7 +669,7 @@ var errEndList = errors.New("end list")
 //
 // if findFile is set it will look for files called (bucket, directory)
-func (f *Fs) list(ctx context.Context, bucket, directory, prefix string, addBucket bool, recurse bool, limit int, hidden bool, findFile bool, fn listFn) error {
+func (f *Fs) list(ctx context.Context, bucket, directory, prefix string, addBucket bool, recurse bool, limit int, hidden bool, versionAt string, findFile bool, fn listFn) error {
```
I'd say don't change the arguments for the function since the desired item is in `f` anyway.

ncw commented on backend/b2/b2.go:
```diff
@@ -1828,6 +1860,9 @@ func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, op
 	if o.fs.opt.Versions {
 		return errNotWithVersions
 	}
+	if o.fs.opt.VersionAt != "" {
```
I would set `opt.Versions` in `NewFs` if `VersionAt` is specified. Then we can do away with all these extra checks.

swazrgb replied:
I'm worried that might give incorrect behavior in certain obscure situations. When `opt.Versions` is set, various methods change their behavior because they expect filenames to have the version suffixes, which is not the case in version-at mode.

`getMetaDataListing` will try to strip the version suffix if `opt.Versions` is set, so if a regular file with such a version suffix happens to be present, unexpected behavior will occur. Instead it should call list as normal, which will return the correct result.

ncw replied:
OK fair enough!

ncw commented on backend/b2/b2.go:
```diff
@@ -106,6 +107,11 @@ in the [b2 integrations checklist](https://www.backblaze.com/b2/docs/integration
+		}, {
+			Name:     "version_at",
+			Help:     "Show file versions as they were at the specified time.\n\nNote that when using this no file write operations are permitted,\nso you can't upload files or delete them.",
+			Default:  "",
```
I think using a `fs.Duration` for the type would be best here. You'd then be able to say

```
--version-at 1w # for 1 week ago
--version-at 2006-01-01 # for 1st jan 2006
```

Or any of the formats here

```
        time.RFC3339,
        "2006-01-02T15:04:05",
        "2006-01-02 15:04:05",
        "2006-01-02",
```

This wants to be converted into a `time.Time` in `NewFs` to make it absolute.

swazrgb replied:
That looks way nicer. What is the best way to convert this duration to an absolute time? I'm thinking:

```
	var versionAtTime time.Time
	if opt.VersionAt.IsSet() {
		versionAtTime = time.Now().Add(-time.Duration(opt.VersionAt))
	}
	f := &Fs{
		...,
		versionAtTime: versionAtTime,
	}
```

But there is no guarantee that `time.Now()` is the same value here as it was when the duration was parsed, which matters if the user specifies an absolute time using `--b2-version-at=2022-03-13T00:00:00Z`. I'd expect a `time.Time` of `2022-03-15 04:47:56.0`, not something like `2022-03-15 04:47:56.10338005`.

The `ParseDuration` appears to offset from `time.Now()`, instead of some accessible constant. So, it's close, but not exactly the same as the user input. Probably not a huge deal, especially since the b2 api counts time by the second, but I am concerned this may not be correct.

Is this fine, or do you have a better suggestion?

swazrgb replied:
Thanks for the suggestion. I pushed 2 new commits, one implementing the new `fs.Time`, and one to use it in b2. Now the parsing step in `newFs` is no longer necessary either.

Could you please let me know if this looks ok?

ncw commented on backend/b2/b2.go:
```diff
@@ -211,6 +217,7 @@ type Options struct {
 	Endpoint                      string               `config:"endpoint"`
 	TestMode                      string               `config:"test_mode"`
 	Versions                      bool                 `config:"versions"`
+	VersionAt                     string               `config:"version_at"`
```
`fs.Duration` here

ncw reviewed (commented):
Nice feature - thank you :-)

See inline for comments

swazrgb commented on fs/parsetime.go:
```diff
@@ -0,0 +1,106 @@
+
+// For overriding in unittests.
+var (
+	timeNowFunc = func() time.Time { return time.Now() }
```
I had to add this mock so the `TestParseTimeUnmarshalJSON` & `TestTimeScan` could test duration offsets as well

ncw replied:
This should just be

`timeNowFunc = time.Now`

Awbmilne commented:
Oh my god. You're an absolute hero.!
This is so awesome!
BackBlaze B2 Backups, HERE I COME!

ncw commented on fs/parsetime.go:
```diff
@@ -0,0 +1,106 @@
+func parseTimeDates(date string) (t time.Time, err error) {
+	var instant time.Time
+	for _, timeFormat := range timeFormats {
+		instant, err = time.Parse(timeFormat, date)
```
This should be `time.ParseInLocation(timeFormat, date, time.Local)`

See the very recently fixed 887cccb2c17fa5f9c70ce32bc4c949e5906897a7

In fact you you'll want to update your tests based on that too.

I think you should probably get `parseDurationDates` to call this function so we don't duplicate the logic! I'd move this function to `parseduration.go` to be with the list of times also.

ncw commented on fs/parsetime.go:
```diff
@@ -0,0 +1,106 @@
+
+	now := timeNowFunc()
+
+	// Attempt to parse as unix timestamp
+	unixMillis, err := strconv.ParseInt(date, 10, 64)
+	if err == nil && unixMillis >= 0 {
+		return time.UnixMilli(unixMillis).UTC(), nil
+	}
```
Do we really need this? It isn't a good date format!

swazrgb replied:
The main reason I added this is because of the JSON unmarshalling being there, and whenever I encounter timestamps in JSON they are unix millis (`Date.now()` in javascript). But I'm not sure why the json unmarshalling is there, but I copied it from `fs.Duration`. Either way I'd be fine removing it, I don't see a real need for it otherwise.

How should I implement `UnmarshalJSONFlag`'s `setInt` method if not a unix milli? Just raise an error?

ncw replied:
Ah I see. The JSON parsing of fs.Duration as an integer is because it is returned as an integer number of Nanoseconds so needs to be parsed that way too. I think we can do without the integer parsing here, so I think we should remove this.

swazrgb replied:
Alright, I've removed UnixMilli parsing in https://github.com/rclone/rclone/pull/6035/commits/e57cebdacb86ab2866e81811cad9458690bd2ac4

ncw reviewed (commented):
I think this is looking great!

I put a couple of small things to fixup inline.

Can you rebase on master, squash into two commits

1. addition of fs.Time
2. Changes to b2

Then force push please?

Hopefully that will be it :-)

swazrgb commented:
@ncw Done! I did leave the unix milli parsing in there for now, but please let me know what you think about https://github.com/rclone/rclone/pull/6035#discussion_r840980866

swazrgb commented:
@ncw Could you take another look? Thanks!

ncw reviewed (commented):
Sorry for the delay.

I think the CI shows a broken test and I left one comment inline about the `int` parsing.

If you can fix those up, rebase on master then I'll merge - thank you :-)

swazrgb commented:
@ncw No problem, thanks for the review! I've removed the `int` parsing in https://github.com/rclone/rclone/pull/6035/commits/e57cebdacb86ab2866e81811cad9458690bd2ac4

The test failure was dependent on the timezone of the machine running the tests. My machine is running `Europe/Amsterdam`, and CI is using `UTC` presumably. This affects the result of `time.ParseInLocation` when `time.Local` is used as you indicated in https://github.com/rclone/rclone/pull/6035#discussion_r840501255

On my machine with `TZ=Europe/Amsterdam` this returns `time.FixedZone("", 0)`, but when ran with `TZ=UTC` this returns `time.Local`. To make the test timezone independent I've switched to `time.Equal` instead in https://github.com/rclone/rclone/pull/6035/commits/bfa33026a3fd4ea4c7935fd6a6f6bf54982f4c48

It should pass now :crossed_fingers:

swazrgb commented:
Thanks for running the tests. Looks like I missed removing one UnixMillis test :sweat_smile: 

Removed that test & ran the tests locally on 1.16 as well now.

ncw reviewed (commented):
This is looking great now - thank you :-)

Can you squash it down into two commits, one for implementing `fs.Time` - that should come first, and one for the b2 changes

[fs: Implement fs.Time](https://github.com/rclone/rclone/pull/6035/commits/0051969e496de9d25d30513439ebfe7560167158) 
- [fs: Compare times using time.Equal instead of assert.Equal](https://github.com/rclone/rclone/pull/6035/commits/bfa33026a3fd4ea4c7935fd6a6f6bf54982f4c48) 

[b2: Add b2-version-at flag to show file versions at time](https://github.com/rclone/rclone/pull/6035/commits/de479c63ef7eebcce2d527e613609f938b697c75) 
- [b2: Use fs.Duration and only parse once in newFs](https://github.com/rclone/rclone/pull/6035/commits/f65ac77c17a714cc836db497bf768db493b99bd4)
- [b2: Use fs.Time for version-at](https://github.com/rclone/rclone/pull/6035/commits/7cae03bcf413d334fcede5b723f66a6891c2436d)
- [fs: Remove time.UnixMilli parsing](https://github.com/rclone/rclone/pull/6035/commits/d4a897b2957ef644bb0b1c3bb29545f4537e84af) 

Then I'll merge it.

I can do the above squashing if you would prefer (you can't do it though the github interface alas, so I'd pull it locally).

swazrgb commented:
@ncw Thanks, done!

swazrgb commented:
Nice, thanks! Unfortunately I don't really use s3 myself nor am I familiar with its API. 

While reading that issue I did come across #1532 which I'll mention here, since this PR covers a few of the use cases mentioned there (albeit just for B2), so the issues are linked together.

Awbmilne commented:
Would this work through a crypt layer?  
If I had `Crypt > B2`, Could I mount my Crypt volume using the --b2-version-at flag?  
There is support for using --b2-versions through Crypt in #5205.  
Im unfamiliar with Rclone's internals, so this may be intrinsic to how remote layering works.

ncw commented:
@Awbmilne I think it should - give it a go!
